File: mobile/lib/models/card.dart
// ─── Card Model ──────────────────────────────────────────────

enum Suit { hearts, diamonds, clubs, spades }

enum Rank {
  ace(1), two(2), three(3), four(4), five(5), six(6), seven(7),
  eight(8), nine(9), ten(10), jack(11), queen(12), king(13);

  final int value;
  const Rank(this.value);
}

class Card {
  final int id;
  final Suit? suit;
  final Rank? rank;
  final bool isJoker;
  final int deckIndex;

  const Card({
    required this.id,
    this.suit,
    this.rank,
    this.isJoker = false,
    this.deckIndex = 0,
  });

  String get display {
    if (isJoker) return '🃏';
    final rankStr = {
      Rank.ace: 'A', Rank.two: '2', Rank.three: '3', Rank.four: '4',
      Rank.five: '5', Rank.six: '6', Rank.seven: '7', Rank.eight: '8',
      Rank.nine: '9', Rank.ten: '10', Rank.jack: 'J', Rank.queen: 'Q',
      Rank.king: 'K',
    }[rank] ?? '?';
    final suitStr = {
      Suit.hearts: '♥', Suit.diamonds: '♦', Suit.clubs: '♣', Suit.spades: '♠',
    }[suit] ?? '?';
    return '$rankStr$suitStr';
  }

  bool get isRed => suit == Suit.hearts || suit == Suit.diamonds;

  factory Card.fromJson(Map<String, dynamic> json) {
    // Handle suit — server sends 'hearts', 'joker' etc.
    Suit? suit;
    final suitStr = json['suit']?.toString();
    if (suitStr != null && suitStr != 'joker') {
      suit = Suit.values.where((s) => s.name == suitStr).firstOrNull;
    }

    // Handle rank — server sends string ('A','2','J','JOKER'), local engine sends int
    Rank? rank;
    final rawRank = json['rank'];
    if (rawRank != null) {
      if (rawRank is int) {
        rank = Rank.values.where((r) => r.value == rawRank).firstOrNull;
      } else if (rawRank is String && rawRank != 'JOKER') {
        const rankMap = {
          'A': Rank.ace, '2': Rank.two, '3': Rank.three, '4': Rank.four,
          '5': Rank.five, '6': Rank.six, '7': Rank.seven, '8': Rank.eight,
          '9': Rank.nine, '10': Rank.ten, 'J': Rank.jack, 'Q': Rank.queen, 'K': Rank.king,
        };
        rank = rankMap[rawRank];
      }
    }

    return Card(
      id: json['id'] as int,
      suit: suit,
      rank: rank,
      isJoker: json['isJoker'] as bool? ?? false,
      deckIndex: json['deckIndex'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'suit': suit?.name,
    'rank': rank?.value,
    'isJoker': isJoker,
    'deckIndex': deckIndex,
  };

  @override
  bool operator ==(Object other) => other is Card && other.id == id;

  @override
  int get hashCode => id.hashCode;
}

File: shared/src/types/card.ts
// ─── Card Types ───────────────────────────────────────────────

export enum Suit {
  Hearts = 'hearts',
  Diamonds = 'diamonds',
  Clubs = 'clubs',
  Spades = 'spades',
}

export enum Rank {
  Ace = 1,
  Two = 2,
  Three = 3,
  Four = 4,
  Five = 5,
  Six = 6,
  Seven = 7,
  Eight = 8,
  Nine = 9,
  Ten = 10,
  Jack = 11,
  Queen = 12,
  King = 13,
}

export interface Card {
  /** Unique id within the deck (0..107 for 2×52+jokers) */
  id: number;
  suit: Suit | null;    // null for jokers
  rank: Rank | null;    // null for jokers
  isJoker: boolean;
  /** Which deck copy (0 or 1) */
  deckIndex: number;
}

export const RANK_NAMES: Record<number, string> = {
  1: 'A', 2: '2', 3: '3', 4: '4', 5: '5', 6: '6', 7: '7',
  8: '8', 9: '9', 10: '10', 11: 'J', 12: 'Q', 13: 'K',
};

export const SUIT_SYMBOLS: Record<string, string> = {
  hearts: '♥', diamonds: '♦', clubs: '♣', spades: '♠',
};

export function cardToString(card: Card): string {
  if (card.isJoker) return `🃏(${card.id})`;
  return `${RANK_NAMES[card.rank!]}${SUIT_SYMBOLS[card.suit!]}`;
}

